src/main/scala/com.snowplowanalytics.sauna/utils/Command.scala

```diff
@@ -0,0 +1,226 @@
+              case JsSuccess(command, _) => processCommand[T](command)
+              case JsError(error) => Left(s"Encountered an issue while parsing Sauna command: $error")
+            }
+          case Left(error) => Left(s"Could not validate command JSON: $error")
+        }
+      case JsError(error) => Left(s"Encountered an issue while parsing self-describing JSON: $error")
+    }
+  }
+
+  /**
+   * Processes a Sauna command, validating its' envelope and extracting the data.
+   *
+   * @param command      A Sauna command.
+   * @param igluResolver Optional [[Resolver]] containing info about Iglu repositories.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the processing was successful, Left containing
+   *         an error message otherwise.
+   */
+  def processCommand[T](command: SaunaCommand,
+    igluResolver: Resolver = Resolver(500, List(
+      HttpRepositoryRef(RepositoryRefConfig("Iglu central", 0, List("com.snowplowanalytics")), "http://iglucentral.com")))
+  )(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    validateSelfDescribing(command.envelope, igluResolver) match {
+      case Right(_) =>
+        command.envelope.data.validate[CommandEnvelope] match {
+          case JsSuccess(envelope, _) =>
+            validateSelfDescribing(command.command, igluResolver) match {
+              case Right(_) =>
+                command.command.data.validate[T] match {
+                  case JsSuccess(data, _) => Right((envelope, data))
+                  case JsError(error) => Left(s"Encountered an issue while parsing Sauna command data: $error")
+                }
+              case Left(error) => Left(s"Could not validate command data JSON: $error")
+            }
+          case JsError(error) => Left(s"Encountered an issue while parsing Sauna command envelope: $error")
+        }
+      case Left(error) => Left(s"Could not validate command envelope JSON: $error")
+    }
+  }
+
+  /**
+   * Validates self-describing data against its' schema.
+   *
+   * @param selfDescribing The self-describing data to validate.
+   * @param igluResolver   [[Resolver]] containing info about Iglu repositories.
+   * @return Right if the data was successfully validated against the schema,
+   *         Left with an error message otherwise.
+   */
+  def validateSelfDescribing(selfDescribing: SelfDescribing, igluResolver: Resolver): Either[String, Unit] = {
+    igluResolver.lookupSchema(selfDescribing.schema) match {
+      case Success(schema) =>
+        Json.fromJson[JsonNode](selfDescribing.data) match {
+          case JsSuccess(jsonNodeData, _) =>
+            val jsonSchema = JsonSchemaFactory.byDefault().getJsonSchema(schema)
+            val processingReport = jsonSchema.validate(jsonNodeData)
+            if (processingReport.isSuccess) {
+              Right()
```

Comment:
I think it actually makes sense - `Either[String, Unit]` makes more sense to me as signal of possible error than `Option[String]`, where `Some(value)` is usually success.